src/utils/InputSystem/InputSystem.ts

```diff
@@ -0,0 +1,246 @@
+  GamepadContext,
+  GamepadBtnContext,
+  SpaceMouseContext,
+} from 'utils/InputSystem/@types'
+
+/* TODO
+ * - button axis (RT, LT)
+ * - support key combinations?
+ */
+
+const isInvalidGamepad = (gamepad: Gamepad) =>
+  gamepad.id.includes('Unknown Gamepad') ||
+  (gamepad.axes.length === 0 && gamepad.buttons.length === 0)
+
+const keyboardCtx: Context = { type: 'keyboard' }
+const emptyCtx: Context = { type: 'none' }
+
+const isPressed = (rawBtn: GamepadButton): boolean =>
+  typeof rawBtn == 'number' ? rawBtn > 0.1 : rawBtn.pressed
+
+/**
+ * This class is used to handle multiple input devices at the same time. To use
+ * it you need to give it an array of actions. Each actions can take one or more
+ * bindings that will be checked in a loop.
+ *
+ * If any of those bindings are true it will perform the action.
+ *
+ * When performing an action a Context is passed to the callback. This Context
+ * contains information like the gamepad id or it's index. If you want to support
+ * multiple gamepads at the same time you could check the index when an action is
+ * performed and keep a map of users/index.
+ *
+ * Example usage:
+ *
+ * const inputsys = new InputSystem([
+ *    {
+ *      name: 'estop',
+ *      bindings: [
+ *        { type: 'keyboard', code: 'Space' },
+ *        { type: 'gamepadBtn', button: buttons.A },
+ *      ],
+ *      perform: (ctx) => {
+ *        console.log('performed estop')
+ *      },
+ *    },
+ * ])
+ * inputsys.start()
+ *
+ */
+export class InputSystem {
+  isRunning = false
+  private actions: Action[] = []
+  private lastTimestamp: number[] = []
+  private lastGamepad: Gamepad[] = []
+  private isBrowserSupported = true
+
+  constructor(actions: Action[]) {
+    this.actions = actions
+
+    if (!navigator.getGamepads) {
+      this.isBrowserSupported = false
+      console.warn('This browser does not support gamepads.')
+      return
+    }
+
+    this.initEventListeners()
+  }
+
+  start = () => {
+    if (this.isBrowserSupported && !this.isRunning) {
+      this.isRunning = true
+      this.update()
+    }
+  }
+
+  stop = () => {
+    this.isRunning = false
+  }
+
+  setActionMap = (actions: Action[]) => {
+    this.actions = actions
+  }
+
+  private update = () => {
+    if (this.isRunning) {
+      this.updateGamepad()
+      requestAnimationFrame(this.update)
+    }
+  }
+
+  private checkActionsBindings = (
+    checkBinding: (binding: Binding) => [boolean, Context]
+  ) => {
+    for (const action of this.actions) {
+      for (const binding of action.bindings) {
+        const [canPerform, context] = checkBinding(binding)
+        if (canPerform) action.perform(context)
+      }
+    }
+  }
+
+  private updateGamepad = () => {
+    const gamepads = navigator.getGamepads()
+    for (let i = 0; i <= gamepads.length; i++) {
+      const gamepad = gamepads[i]
+
+      if (!gamepad || isInvalidGamepad(gamepad)) continue
+
+      // if (gamepad.timestamp && gamepad.timestamp === this.lastTimestamp[i])
```

Comment:
Maybe remove this comment ?

Reply:
The thing is, if I never commit that line it will never be part of the history if we ever want to go back to this solution. I'll remove it in a separate commit once this is merged

Reply:
Okay, I wasn't sure what was that supposed to be. I guess we could do what you suggest. Although wouldn't there be previous commits in the master branch with that code since it's the older system?

Reply:
No, it wasn't exactly like that in the old system. This is specific to the new system. It's simple enough that I'll just remove it. We don't need it anyway.